File: physicar_deepracer/physicar_deepracer/inference_engine.py
"""
TFLite Inference Engine for PhysiCar DeepRacer

Handles:
1. Loading TFLite model
2. Image preprocessing (grayscale, resize)
3. LiDAR data preprocessing
4. Running inference
5. Action selection from probabilities
"""

import os
import numpy as np
from typing import Optional, Tuple, List, Dict, Any
from dataclasses import dataclass, field
from enum import Enum

# Pre-import tflite runtime at module level to avoid cold start on first model load
try:
    import tflite_runtime.interpreter as _tflite
except ImportError:
    try:
        import tensorflow.lite as _tflite
    except ImportError:
        _tflite = None

from . import constants
from .model_loader import ModelInfo, ActionSpace


class ActionSelectionMode(Enum):
    """Action selection strategy for discrete action space"""
    GREEDY = "greedy"        # Always select highest probability action
    STOCHASTIC = "stochastic"  # Sample from probability distribution
    MEAN = "mean"            # Weighted average of all actions by probability


@dataclass
class InferenceResult:
    """Result of a single inference"""
    action_index: int
    action: ActionSpace
    probabilities: np.ndarray
    speed: float           # m/s
    steering_angle: float  # degrees (negative = left, positive = right)


@dataclass
class LoadedModel:
    """A model loaded into memory with its TFLite interpreter"""
    model_info: ModelInfo
    interpreter: Any
    input_details: Any
    output_details: Any


class InferenceEngine:
    """
    TFLite inference engine for DeepRacer models.
    
    Supports multiple models loaded simultaneously.
    One model is "active" for inference at a time.
    Switching between loaded models is instant (no re-loading).
    """
    
    def __init__(self, logger=None, action_selection_mode: ActionSelectionMode = ActionSelectionMode.GREEDY):
        self.logger = logger
        self._loaded_models: Dict[str, LoadedModel] = {}
        self._active_model_name: Optional[str] = None
        
        # Action selection mode
        self._action_selection_mode = action_selection_mode
        
        # Inference statistics
        self._inference_count = 0
        self._last_result: Optional[InferenceResult] = None
    
    def _log(self, msg: str, level: str = "info"):
        if self.logger:
            getattr(self.logger, level)(msg)
        else:
            print(f"[{level.upper()}] {msg}")
    
    def load_model(self, model_info: ModelInfo) -> Tuple[bool, str]:
        """
        Load TFLite model into memory pool and set as active.
        
        If the model is already loaded, just switches active model (instant).
        Multiple models can be loaded simultaneously.
        
        Args:
            model_info: Model information from ModelLoader
            
        Returns:
            Tuple of (success, message)
        """
        name = model_info.name
        
        # Already loaded? Just switch active
        if name in self._loaded_models:
            self._active_model_name = name
            self._log(f"Model '{name}' already loaded, switched to active")
            return True, f"Model '{name}' already loaded, switched to active"
        
        if not os.path.exists(model_info.tflite_path):
            return False, f"TFLite file not found: {model_info.tflite_path}"
        
        try:
            if _tflite is None:
                return False, "tflite_runtime and tensorflow are both unavailable"
            
            tflite = _tflite
            
            # Create interpreter with multi-threading
            num_threads = max(1, os.cpu_count() - 2) if os.cpu_count() else 2
            interpreter = tflite.Interpreter(
                model_path=model_info.tflite_path,
                num_threads=num_threads
            )
            interpreter.allocate_tensors()
            
            # Get input/output details
            input_details = interpreter.get_input_details()
            output_details = interpreter.get_output_details()
            
            # Store loaded model
            self._loaded_models[name] = LoadedModel(
                model_info=model_info,
                interpreter=interpreter,
                input_details=input_details,
                output_details=output_details,
            )
            self._active_model_name = name
            self._inference_count = 0
            
            self._log(f"TFLite model '{name}' loaded with {num_threads} threads")
            self._log(f"Input tensors: {len(input_details)}")
            for i, inp in enumerate(input_details):
                self._log(f"  [{i}] {inp['name']}: {inp['shape']} {inp['dtype']}")
            self._log(f"Output tensors: {len(output_details)}")
            for i, out in enumerate(output_details):
                self._log(f"  [{i}] {out['name']}: {out['shape']} {out['dtype']}")
            self._log(f"Loaded models: {list(self._loaded_models.keys())}")
            
            return True, "Model loaded successfully"
            
        except Exception as e:
            error_msg = f"Failed to load TFLite model: {str(e)}"
            self._log(error_msg, "error")
            return False, error_msg
    
    def unload_model(self, name: str) -> Tuple[bool, str]:
        """
        Unload a model from memory.
        
        Args:
            name: Model name, or empty string to unload all
            
        Returns:
            Tuple of (success, message)
        """
        if not name:
            # Unload all
            count = len(self._loaded_models)
            self._loaded_models.clear()
            self._active_model_name = None
            self._last_result = None
            self._log(f"Unloaded all {count} models")
            return True, f"Unloaded all {count} models"
        
        if name not in self._loaded_models:
            return False, f"Model '{name}' is not loaded"
        
        del self._loaded_models[name]
        
        # If we unloaded the active model, pick another or None
        if self._active_model_name == name:
            if self._loaded_models:
                self._active_model_name = next(iter(self._loaded_models))
                self._log(f"Active model switched to '{self._active_model_name}'")
            else:
                self._active_model_name = None
                self._last_result = None
        
        self._log(f"Model '{name}' unloaded. Loaded: {list(self._loaded_models.keys())}")
        return True, f"Model '{name}' unloaded"
    
    def preprocess_image(self, image: np.ndarray) -> np.ndarray:
        """
        Preprocess camera image for inference
        
        Args:
            image: BGR image from camera (480x320 or similar)
            
        Returns:
            Preprocessed image [1, 120, 160, 1] float32, normalized 0-1
        """
        import cv2
        
        # Convert to grayscale if color
        if len(image.shape) == 3 and image.shape[2] == 3:
            gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        else:
            gray = image
        
        # Resize to inference dimensions
        resized = cv2.resize(
            gray, 
            (constants.INFERENCE_IMAGE_WIDTH, constants.INFERENCE_IMAGE_HEIGHT),
            interpolation=cv2.INTER_AREA
        )
        
        # Keep 0-255 range as float32 (DeepRacer training does NOT normalize)
        img_float = resized.astype(np.float32)
        
        # Shape: [1, height, width, channels]
        return img_float.reshape(
            1, 
            constants.INFERENCE_IMAGE_HEIGHT, 
            constants.INFERENCE_IMAGE_WIDTH, 
            constants.INFERENCE_IMAGE_CHANNELS
        )
    
    def preprocess_lidar(self, ranges: np.ndarray, angle_min: float, 
                         angle_increment: float) -> np.ndarray:
        """
        Preprocess LiDAR scan data for inference
        
        DeepRacer training simulation uses 300° FOV: -150° ~ +150°
        (rear 60° is excluded)
        
        ROS standard coordinate:
        - 0° = front
        - + = left (counter-clockwise)
        - - = right (clockwise)
        
        Output index order (matching training):
        - Index 0 = -150° (right rear)
        - Index 31-32 = 0° (front)
        - Index 63 = +150° (left rear)
        
        Args:
            ranges: Raw range data from LiDAR
            angle_min: Minimum angle (radians)
            angle_increment: Angle between samples (radians)
            
        Returns:
            Interpolated data [1, 64] float32, distances in meters (clipped 0.15-1.0)
        """
        # Target angle range (300° FOV, excluding rear 60°)
        target_min_deg = constants.LIDAR_MIN_ANGLE_DEG  # -150°
        target_max_deg = constants.LIDAR_MAX_ANGLE_DEG  # +150°
        num_values = constants.LIDAR_SECTORS  # 64
        min_dist = constants.LIDAR_MIN_DIST  # 0.15m
        max_dist = constants.LIDAR_MAX_DIST  # 1.0m
        
        # Build arrays of (angle_deg, distance) for samples within target range
        lidar_angles = []
        lidar_ranges = []
        
        for i, distance in enumerate(ranges):
            # Calculate angle in radians, then convert to degrees
            angle_rad = angle_min + i * angle_increment
            # Convert to degrees in range -180 to +180
            angle_deg = np.degrees(angle_rad)
            # Normalize to -180 ~ +180 range
            while angle_deg > 180:
                angle_deg -= 360
            while angle_deg < -180:
                angle_deg += 360
            
            # Check if within target range: -150° to +150°
            # This excludes the rear 60° (from -180° to -150° and +150° to +180°)
            if target_min_deg - 0.5 <= angle_deg <= target_max_deg + 0.5:
                # Clip distance, handle inf as max_dist
                if not np.isfinite(distance) or distance <= 0:
                    clipped_dist = max_dist
                else:
                    clipped_dist = np.clip(distance, min_dist, max_dist)
                
                lidar_angles.append(angle_deg)
                lidar_ranges.append(clipped_dist)
        
        # If no samples in range, return max distance array
        if len(lidar_angles) < 2:
            return np.full((1, num_values), max_dist, dtype=np.float32)
        
        # Sort by angle (ascending: -150° → +150°)
        sorted_indices = np.argsort(lidar_angles)
        lidar_angles = np.array(lidar_angles)[sorted_indices]
        lidar_ranges = np.array(lidar_ranges)[sorted_indices]
        
        # Ensure boundary angles exist for interpolation
        if lidar_angles[0] > target_min_deg:
            lidar_angles = np.insert(lidar_angles, 0, target_min_deg)
            lidar_ranges = np.insert(lidar_ranges, 0, max_dist)
        if lidar_angles[-1] < target_max_deg:
            lidar_angles = np.append(lidar_angles, target_max_deg)
            lidar_ranges = np.append(lidar_ranges, max_dist)
        
        # Generate target angles (evenly spaced from -150° to +150°)
        # Index 0 = -150° (right rear), Index 63 = +150° (left rear)
        target_angles = np.linspace(target_min_deg, target_max_deg, num_values)
        
        # Interpolate to get 64 evenly-spaced values
        interpolated = np.interp(target_angles, lidar_angles, lidar_ranges)
        
        # Shape: [1, num_values]
        return interpolated.astype(np.float32).reshape(1, num_values)
    
    def run_inference(self, image: np.ndarray, lidar: Optional[np.ndarray] = None) -> Optional[InferenceResult]:
        """
        Run inference on preprocessed inputs.
        
        Input tensors are matched by name (deepracer-custom-car pattern):
          - 'FRONT_FACING_CAMERA' or 'observation' → image
          - 'LIDAR' → lidar (skipped if model has no LiDAR tensor)
        
        Args:
            image: Preprocessed image [1, 120, 160, 1]
            lidar: Preprocessed LiDAR [1, 64] or None for camera-only models
            
        Returns:
            InferenceResult or None on error
        """
        if self._active_model_name is None or self._active_model_name not in self._loaded_models:
            self._log("No model loaded", "error")
            return None
        
        active = self._loaded_models[self._active_model_name]
        
        try:
            # Set inputs by tensor name dispatch (handles any sensor combo)
            for inp in active.input_details:
                name = inp['name'].lower()
                if 'camera' in name or 'observation' in name:
                    active.interpreter.set_tensor(inp['index'], image)
                elif 'lidar' in name:
                    if lidar is not None:
                        active.interpreter.set_tensor(inp['index'], lidar)
            
            # Run inference
            active.interpreter.invoke()
            
            # Get output probabilities
            output = active.interpreter.get_tensor(active.output_details[0]['index'])
            probabilities = output.flatten()
            
            # Convert to proper probability distribution (softmax if needed)
            # TFLite PPO output is already softmax, but ensure non-negative and sum to 1
            probs = np.clip(probabilities, 0, None)
            prob_sum = probs.sum()
            if prob_sum > 0:
                probs = probs / prob_sum
            else:
                probs = np.ones_like(probs) / len(probs)
            
            model_info = active.model_info
            
            # Select action based on mode
            if self._action_selection_mode == ActionSelectionMode.GREEDY:
                # Always select highest probability action
                action_idx = int(np.argmax(probs))
                action = model_info.action_space[action_idx]
                speed = action.speed
                steering_angle = action.steering_angle
            elif self._action_selection_mode == ActionSelectionMode.MEAN:
                # Weighted average of all actions by probability
                speed = sum(probs[i] * model_info.action_space[i].speed 
                           for i in range(len(probs)))
                steering_angle = sum(probs[i] * model_info.action_space[i].steering_angle 
                                    for i in range(len(probs)))
                action_idx = int(np.argmax(probs))  # Report highest prob action for logging
                action = model_info.action_space[action_idx]
            else:  # STOCHASTIC
                # Sample from probability distribution
                action_idx = int(np.random.choice(len(probs), p=probs))
                action = model_info.action_space[action_idx]
                speed = action.speed
                steering_angle = action.steering_angle
            
            result = InferenceResult(
                action_index=action_idx,
                action=action,
                probabilities=probabilities,
                speed=speed,
                steering_angle=steering_angle
            )
            
            self._inference_count += 1
            self._last_result = result
            
            return result
            
        except Exception as e:
            self._log(f"Inference error: {str(e)}", "error")
            return None
    
    def infer_from_raw(self, image: np.ndarray, ranges: Optional[np.ndarray] = None,
                       angle_min: float = 0.0, angle_increment: float = 0.0) -> Optional[InferenceResult]:
        """
        Convenience method: preprocess and run inference in one call.
        
        Args:
            image: Raw BGR camera image
            ranges: Raw LiDAR range data (None for camera-only models)
            angle_min: LiDAR minimum angle
            angle_increment: LiDAR angle increment
            
        Returns:
            InferenceResult or None
        """
        proc_image = self.preprocess_image(image)
        proc_lidar = None
        if ranges is not None:
            proc_lidar = self.preprocess_lidar(ranges, angle_min, angle_increment)
        return self.run_inference(proc_image, proc_lidar)
    
    @property
    def inference_count(self) -> int:
        return self._inference_count
    
    @property
    def last_result(self) -> Optional[InferenceResult]:
        return self._last_result
    
    @property
    def is_loaded(self) -> bool:
        return self._active_model_name is not None and self._active_model_name in self._loaded_models
    
    @property
    def active_model_name(self) -> Optional[str]:
        return self._active_model_name
    
    @property
    def active_model_info(self) -> Optional[ModelInfo]:
        if self._active_model_name and self._active_model_name in self._loaded_models:
            return self._loaded_models[self._active_model_name].model_info
        return None
    
    @property
    def loaded_model_names(self) -> List[str]:
        return list(self._loaded_models.keys())
    
    @property
    def action_selection_mode(self) -> ActionSelectionMode:
        return self._action_selection_mode
    
    @action_selection_mode.setter
    def action_selection_mode(self, mode: ActionSelectionMode):
        self._action_selection_mode = mode
        if self.logger:
            self.logger.info(f"Action selection mode set to: {mode.value}")

File: physicar_deepracer/src/inference_engine.cpp
#include "physicar_deepracer/inference_engine.hpp"
#include "physicar_deepracer/constants.hpp"

#include <algorithm>
#include <cmath>
#include <cstring>
#include <numeric>
#include <thread>

#include <opencv2/imgproc.hpp>
#include <rclcpp/logging.hpp>
#include <tensorflow/lite/kernels/register.h>

namespace physicar_deepracer {

InferenceEngine::InferenceEngine(rclcpp::Logger logger, ActionSelectionMode mode)
    : logger_(logger),
      action_selection_mode_(mode),
      rng_(std::random_device{}()) {}

std::pair<bool, std::string>
InferenceEngine::load_model(std::shared_ptr<ModelInfo> model_info) {
    const auto& name = model_info->name;

    // Already loaded? Just switch active
    if (loaded_models_.count(name)) {
        active_model_name_ = name;
        RCLCPP_INFO(logger_, "Model '%s' already loaded, switched to active", name.c_str());
        return {true, "Model '" + name + "' already loaded, switched to active"};
    }

    try {
        auto loaded = std::make_unique<LoadedModel>();
        loaded->model_info = model_info;

        // Load FlatBuffer model
        loaded->flat_buffer_model =
            tflite::FlatBufferModel::BuildFromFile(model_info->tflite_path.c_str());
        if (!loaded->flat_buffer_model) {
            return {false, "Failed to load TFLite file: " + model_info->tflite_path};
        }

        // Build interpreter
        tflite::ops::builtin::BuiltinOpResolver resolver;
        tflite::InterpreterBuilder builder(*loaded->flat_buffer_model, resolver);

        int num_threads = std::max(1,
            static_cast<int>(std::thread::hardware_concurrency()) - 2);
        builder.SetNumThreads(num_threads);
        builder(&loaded->interpreter);

        if (!loaded->interpreter) {
            return {false, "Failed to build TFLite interpreter"};
        }

        if (loaded->interpreter->AllocateTensors() != kTfLiteOk) {
            return {false, "Failed to allocate TFLite tensors"};
        }

        // Log input/output tensor details
        const auto& inputs = loaded->interpreter->inputs();
        RCLCPP_INFO(logger_, "TFLite model '%s' loaded with %d threads",
                    name.c_str(), num_threads);
        RCLCPP_INFO(logger_, "Input tensors: %zu", inputs.size());
        for (size_t i = 0; i < inputs.size(); ++i) {
            const auto* tensor = loaded->interpreter->tensor(inputs[i]);
            std::string shape_str = "[";
            for (int d = 0; d < tensor->dims->size; ++d) {
                if (d > 0) shape_str += ", ";
                shape_str += std::to_string(tensor->dims->data[d]);
            }
            shape_str += "]";
            RCLCPP_INFO(logger_, "  [%zu] %s: %s", i, tensor->name, shape_str.c_str());
        }

        const auto& outputs = loaded->interpreter->outputs();
        RCLCPP_INFO(logger_, "Output tensors: %zu", outputs.size());
        for (size_t i = 0; i < outputs.size(); ++i) {
            const auto* tensor = loaded->interpreter->tensor(outputs[i]);
            std::string shape_str = "[";
            for (int d = 0; d < tensor->dims->size; ++d) {
                if (d > 0) shape_str += ", ";
                shape_str += std::to_string(tensor->dims->data[d]);
            }
            shape_str += "]";
            RCLCPP_INFO(logger_, "  [%zu] %s: %s", i, tensor->name, shape_str.c_str());
        }

        active_model_name_ = name;
        inference_count_ = 0;

        // Log all loaded models
        loaded_models_[name] = std::move(loaded);
        std::string loaded_str;
        for (const auto& [n, _] : loaded_models_) {
            if (!loaded_str.empty()) loaded_str += ", ";
            loaded_str += n;
        }
        RCLCPP_INFO(logger_, "Loaded models: [%s]", loaded_str.c_str());

        return {true, "Model loaded successfully"};

    } catch (const std::exception& e) {
        return {false, std::string("Failed to load TFLite model: ") + e.what()};
    }
}

std::pair<bool, std::string>
InferenceEngine::unload_model(const std::string& name) {
    if (name.empty()) {
        // Unload all
        int count = static_cast<int>(loaded_models_.size());
        loaded_models_.clear();
        active_model_name_.clear();
        last_result_.reset();
        RCLCPP_INFO(logger_, "Unloaded all %d models", count);
        return {true, "Unloaded all " + std::to_string(count) + " models"};
    }

    if (!loaded_models_.count(name)) {
        return {false, "Model '" + name + "' is not loaded"};
    }

    loaded_models_.erase(name);

    // If we unloaded the active model, pick another or none
    if (active_model_name_ == name) {
        if (!loaded_models_.empty()) {
            active_model_name_ = loaded_models_.begin()->first;
            RCLCPP_INFO(logger_, "Active model switched to '%s'",
                        active_model_name_.c_str());
        } else {
            active_model_name_.clear();
            last_result_.reset();
        }
    }

    std::string loaded_str;
    for (const auto& [n, _] : loaded_models_) {
        if (!loaded_str.empty()) loaded_str += ", ";
        loaded_str += n;
    }
    RCLCPP_INFO(logger_, "Model '%s' unloaded. Loaded: [%s]",
                name.c_str(), loaded_str.c_str());
    return {true, "Model '" + name + "' unloaded"};
}

std::vector<float> InferenceEngine::preprocess_image(const cv::Mat& image) const {
    cv::Mat gray;
    if (image.channels() == 3) {
        cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);
    } else {
        gray = image;
    }

    cv::Mat resized;
    cv::resize(gray, resized,
               cv::Size(INFERENCE_IMAGE_WIDTH, INFERENCE_IMAGE_HEIGHT),
               0, 0, cv::INTER_AREA);

    // Keep 0-255 range as float32 (DeepRacer training does NOT normalize)
    cv::Mat float_img;
    resized.convertTo(float_img, CV_32F);

    // Flatten to [1, H, W, 1] contiguous array
    std::vector<float> result(
        INFERENCE_IMAGE_HEIGHT * INFERENCE_IMAGE_WIDTH * INFERENCE_IMAGE_CHANNELS);
    std::memcpy(result.data(), float_img.data, result.size() * sizeof(float));
    return result;
}

std::vector<float> InferenceEngine::preprocess_lidar(
    const std::vector<float>& ranges,
    float angle_min, float angle_increment) const {

    constexpr double target_min_deg = LIDAR_MIN_ANGLE_DEG;  // -150
    constexpr double target_max_deg = LIDAR_MAX_ANGLE_DEG;  // +150
    constexpr int num_values = LIDAR_SECTORS;               // 64
    constexpr double min_dist = LIDAR_MIN_DIST;             // 0.15
    constexpr double max_dist = LIDAR_MAX_DIST;             // 1.0

    // Build arrays of (angle_deg, distance) for samples within target range
    std::vector<double> lidar_angles;
    std::vector<double> lidar_ranges;

    for (size_t i = 0; i < ranges.size(); ++i) {
        double angle_rad = angle_min + static_cast<double>(i) * angle_increment;
        double angle_deg = angle_rad * 180.0 / M_PI;
        // Normalize to -180 ~ +180
        while (angle_deg > 180.0) angle_deg -= 360.0;
        while (angle_deg < -180.0) angle_deg += 360.0;

        // Check if within target range: -150 to +150
        if (angle_deg >= target_min_deg - 0.5 && angle_deg <= target_max_deg + 0.5) {
            double dist = static_cast<double>(ranges[i]);
            double clipped_dist;
            if (!std::isfinite(dist) || dist <= 0.0) {
                clipped_dist = max_dist;
            } else {
                clipped_dist = std::clamp(dist, min_dist, max_dist);
            }
            lidar_angles.push_back(angle_deg);
            lidar_ranges.push_back(clipped_dist);
        }
    }

    // If no samples in range, return max distance array
    if (lidar_angles.size() < 2) {
        return std::vector<float>(num_values, static_cast<float>(max_dist));
    }

    // Sort by angle ascending
    std::vector<size_t> indices(lidar_angles.size());
    std::iota(indices.begin(), indices.end(), 0);
    std::sort(indices.begin(), indices.end(),
              [&](size_t a, size_t b) { return lidar_angles[a] < lidar_angles[b]; });

    std::vector<double> sorted_angles(lidar_angles.size());
    std::vector<double> sorted_ranges(lidar_angles.size());
    for (size_t i = 0; i < indices.size(); ++i) {
        sorted_angles[i] = lidar_angles[indices[i]];
        sorted_ranges[i] = lidar_ranges[indices[i]];
    }

    // Ensure boundary angles exist for interpolation
    if (sorted_angles.front() > target_min_deg) {
        sorted_angles.insert(sorted_angles.begin(), target_min_deg);
        sorted_ranges.insert(sorted_ranges.begin(), max_dist);
    }
    if (sorted_angles.back() < target_max_deg) {
        sorted_angles.push_back(target_max_deg);
        sorted_ranges.push_back(max_dist);
    }

    // Generate target angles (evenly spaced from -150 to +150)
    std::vector<double> target_angles(num_values);
    for (int i = 0; i < num_values; ++i) {
        target_angles[i] = target_min_deg +
            static_cast<double>(i) * (target_max_deg - target_min_deg) /
            static_cast<double>(num_values - 1);
    }

    // Linear interpolation (np.interp equivalent)
    std::vector<float> result(num_values);
    for (int i = 0; i < num_values; ++i) {
        double x = target_angles[i];

        // Find bracketing indices in sorted_angles
        if (x <= sorted_angles.front()) {
            result[i] = static_cast<float>(sorted_ranges.front());
        } else if (x >= sorted_angles.back()) {
            result[i] = static_cast<float>(sorted_ranges.back());
        } else {
            // Binary search for the right bracket
            auto it = std::lower_bound(sorted_angles.begin(), sorted_angles.end(), x);
            size_t idx = static_cast<size_t>(std::distance(sorted_angles.begin(), it));
            if (idx == 0) idx = 1;
            double x0 = sorted_angles[idx - 1];
            double x1 = sorted_angles[idx];
            double y0 = sorted_ranges[idx - 1];
            double y1 = sorted_ranges[idx];
            double t = (x1 != x0) ? (x - x0) / (x1 - x0) : 0.0;
            result[i] = static_cast<float>(y0 + t * (y1 - y0));
        }
    }

    return result;
}

std::optional<InferenceResult> InferenceEngine::run_inference(
    const std::vector<float>& image_input,
    const std::vector<float>* lidar_input) {

    if (active_model_name_.empty() || !loaded_models_.count(active_model_name_)) {
        RCLCPP_ERROR(logger_, "No model loaded");
        return std::nullopt;
    }

    auto& active = loaded_models_[active_model_name_];
    auto* interpreter = active->interpreter.get();

    try {
        // Set inputs by tensor name dispatch
        const auto& inputs = interpreter->inputs();
        for (size_t i = 0; i < inputs.size(); ++i) {
            const auto* tensor = interpreter->tensor(inputs[i]);
            std::string name(tensor->name);
            std::transform(name.begin(), name.end(), name.begin(), ::tolower);

            if (name.find("camera") != std::string::npos ||
                name.find("observation") != std::string::npos) {
                float* input_data = interpreter->typed_tensor<float>(inputs[i]);
                std::memcpy(input_data, image_input.data(),
                           image_input.size() * sizeof(float));
            } else if (name.find("lidar") != std::string::npos) {
                if (lidar_input) {
                    float* input_data = interpreter->typed_tensor<float>(inputs[i]);
                    std::memcpy(input_data, lidar_input->data(),
                               lidar_input->size() * sizeof(float));
                }
            }
        }

        // Run inference
        if (interpreter->Invoke() != kTfLiteOk) {
            RCLCPP_ERROR(logger_, "TFLite Invoke failed");
            return std::nullopt;
        }

        // Get output probabilities
        int output_idx = interpreter->outputs()[0];
        const auto* output_tensor = interpreter->tensor(output_idx);
        const float* output_data = interpreter->typed_tensor<float>(output_idx);
        int output_size = 1;
        for (int d = 0; d < output_tensor->dims->size; ++d) {
            output_size *= output_tensor->dims->data[d];
        }

        std::vector<float> probabilities(output_data, output_data + output_size);

        // Convert to proper probability distribution
        // TFLite PPO output is already softmax, but ensure non-negative and sum to 1
        std::vector<float> probs(probabilities.size());
        for (size_t i = 0; i < probabilities.size(); ++i) {
            probs[i] = std::max(0.0f, probabilities[i]);
        }
        float prob_sum = std::accumulate(probs.begin(), probs.end(), 0.0f);
        if (prob_sum > 0.0f) {
            for (auto& p : probs) p /= prob_sum;
        } else {
            float uniform = 1.0f / static_cast<float>(probs.size());
            std::fill(probs.begin(), probs.end(), uniform);
        }

        const auto& model_info = active->model_info;
        const auto& action_space = model_info->action_space;

        InferenceResult result;
        result.probabilities = probabilities;

        if (action_selection_mode_ == ActionSelectionMode::GREEDY) {
            result.action_index = static_cast<int>(
                std::distance(probs.begin(),
                             std::max_element(probs.begin(), probs.end())));
            result.action = action_space[result.action_index];
            result.speed = result.action.speed;
            result.steering_angle = result.action.steering_angle;

        } else if (action_selection_mode_ == ActionSelectionMode::MEAN) {
            double speed = 0.0, steering = 0.0;
            for (size_t i = 0; i < probs.size(); ++i) {
                speed += probs[i] * action_space[i].speed;
                steering += probs[i] * action_space[i].steering_angle;
            }
            result.speed = speed;
            result.steering_angle = steering;
            result.action_index = static_cast<int>(
                std::distance(probs.begin(),
                             std::max_element(probs.begin(), probs.end())));
            result.action = action_space[result.action_index];

        } else {  // STOCHASTIC
            std::discrete_distribution<int> dist(probs.begin(), probs.end());
            result.action_index = dist(rng_);
            result.action = action_space[result.action_index];
            result.speed = result.action.speed;
            result.steering_angle = result.action.steering_angle;
        }

        inference_count_++;
        last_result_ = result;
        return result;

    } catch (const std::exception& e) {
        RCLCPP_ERROR(logger_, "Inference error: %s", e.what());
        return std::nullopt;
    }
}

std::optional<InferenceResult> InferenceEngine::infer_from_raw(
    const cv::Mat& image,
    const std::vector<float>* ranges,
    float angle_min, float angle_increment) {

    auto proc_image = preprocess_image(image);
    if (ranges) {
        auto proc_lidar = preprocess_lidar(*ranges, angle_min, angle_increment);
        return run_inference(proc_image, &proc_lidar);
    }
    return run_inference(proc_image);
}

bool InferenceEngine::is_loaded() const {
    return !active_model_name_.empty() && loaded_models_.count(active_model_name_);
}

std::shared_ptr<ModelInfo> InferenceEngine::active_model_info() const {
    if (!active_model_name_.empty() && loaded_models_.count(active_model_name_)) {
        return loaded_models_.at(active_model_name_)->model_info;
    }
    return nullptr;
}

std::vector<std::string> InferenceEngine::loaded_model_names() const {
    std::vector<std::string> names;
    for (const auto& [name, _] : loaded_models_) {
        names.push_back(name);
    }
    return names;
}

void InferenceEngine::set_action_selection_mode(ActionSelectionMode mode) {
    action_selection_mode_ = mode;
    const char* mode_str = "greedy";
    if (mode == ActionSelectionMode::STOCHASTIC) mode_str = "stochastic";
    else if (mode == ActionSelectionMode::MEAN) mode_str = "mean";
    RCLCPP_INFO(logger_, "Action selection mode set to: %s", mode_str);
}

}  // namespace physicar_deepracer
